File: xposed-stub/src/main/java/de/robv/android/xposed/XposedBridge.java
package de.robv.android.xposed;

import java.lang.reflect.Member;
import java.util.Collections;
import java.util.Set;

public final class XposedBridge {

    public static int XPOSED_BRIDGE_VERSION = 0;

    private XposedBridge() {
    }

    public static void log(String text) {
        // no-op in compile-time stub
    }

    public static void log(Throwable throwable) {
        // no-op in compile-time stub
    }

    public static XC_MethodHook.Unhook hookMethod(Member hookMethod, XC_MethodHook callback) {
        if (callback == null) {
            throw new IllegalArgumentException("callback must not be null");
        }
        return callback.new Unhook(hookMethod);
    }

    public static Set<XC_MethodHook.Unhook> hookAllMethods(
        Class<?> hookClass,
        String methodName,
        XC_MethodHook callback
    ) {
        return Collections.emptySet();
    }

    public static Set<XC_MethodHook.Unhook> hookAllConstructors(Class<?> hookClass, XC_MethodHook callback) {
        return Collections.emptySet();
    }
}

File: app/src/main/java/io/github/magisk317/relay/xp/compat/XposedBridge.kt
package io.github.magisk317.relay.xp.compat

import java.lang.reflect.Member
import java.lang.reflect.Method
import java.lang.reflect.Modifier

object XposedBridge {

    @JvmStatic
    var XPOSED_BRIDGE_VERSION: Int = 0

    @JvmStatic
    fun getXposedVersion(): Int = XPOSED_BRIDGE_VERSION

    @JvmStatic
    fun log(text: String) {
        XposedRuntime.log(text)
    }

    @JvmStatic
    fun log(throwable: Throwable) {
        XposedRuntime.log(throwable)
    }

    @JvmStatic
    fun hookMethod(hookMethod: Member?, callback: XC_MethodHook): XC_MethodHook.Unhook {
        requireNotNull(hookMethod) { "hookMethod is null" }
        return XposedRuntime.hook(hookMethod, callback)
    }

    @JvmStatic
    fun hookAllMethods(
        hookClass: Class<*>?,
        methodName: String,
        callback: XC_MethodHook,
    ): Set<XC_MethodHook.Unhook> {
        requireNotNull(hookClass) { "hookClass is null" }
        return hookClass.declaredMethods
            .asSequence()
            .filter { method ->
                method.name == methodName && !Modifier.isAbstract(method.modifiers) && !Modifier.isNative(method.modifiers)
            }
            .map { hookMethod(it, callback) }
            .toSet()
    }

    @JvmStatic
    fun hookAllConstructors(
        hookClass: Class<*>?,
        callback: XC_MethodHook,
    ): Set<XC_MethodHook.Unhook> {
        requireNotNull(hookClass) { "hookClass is null" }
        return hookClass.declaredConstructors
            .map { constructor -> hookMethod(constructor, callback) }
            .toSet()
    }
}
